orttraining/orttraining/core/optimizer/shape_optimizer.cc

```diff
@@ -0,0 +1,355 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <limits>
+
+#include "core/common/inlined_containers.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+
+#include "orttraining/core/optimizer/shape_optimizer.h"
+
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+
+// Uncomment to log debug info for SO(Shape Optimizer).
+// #define NEED_SO_LOG_DEBUG_INFO 1
+
+#ifndef SO_LOG_DEBUG_INFO
+#ifdef NEED_SO_LOG_DEBUG_INFO
+#define SO_LOG_DEBUG_INFO(logger, message) LOGS(logger, WARNING) << message
+#else
+#define SO_LOG_DEBUG_INFO(logger, message) \
+  ORT_UNUSED_PARAMETER(logger);            \
+  do {                                     \
+  } while (0)
+#endif
+#endif
```

Comment:
We should try to come up with a better way to analyze logs from a specific class/file instead of defining these macro in every class/file.
We can discuss this more and address it later in a separate PR.